tests/pubsub/testgossipmembership.nim

```diff
@@ -0,0 +1,169 @@
+# Nim-LibP2P
+# Copyright (c) 2023-2024 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+{.used.}
+
+import std/[options, deques, sequtils, enumerate, algorithm, sets]
+import stew/byteutils
+import ../../libp2p/builders
+import ../../libp2p/errors
+import ../../libp2p/crypto/crypto
+import ../../libp2p/stream/bufferstream
+import ../../libp2p/protocols/pubsub/[pubsub, gossipsub, mcache, mcache, peertable]
+import ../../libp2p/protocols/pubsub/rpc/[message, messages]
+import ../../libp2p/switch
+import ../../libp2p/muxers/muxer
+import ../../libp2p/protocols/pubsub/rpc/protobuf
+import utils
+import chronos
+
+import ../helpers
+
+proc noop(data: seq[byte]) {.async: (raises: [CancelledError, LPStreamError]).} =
+  discard
+
+const MsgIdSuccess = "msg id gen success"
+
+suite "GossipSub Topic Membership Tests":
+  teardown:
+    checkTrackers()
+
+  # Addition of Designed Test cases for 6. Topic Membership Tests: https://www.notion.so/Gossipsub-651e02d4d7894bb2ac1e4edb55f3192d
+
+  # Generalized setup function to initialize one or more topics
+  proc setupGossipSub(
+      topics: seq[string], numPeers: int
+  ): (TestGossipSub, seq[Connection]) =
+    let gossipSub = TestGossipSub.init(newStandardSwitch())
+    var conns = newSeq[Connection]()
+
+    for topic in topics:
+      gossipSub.mesh[topic] = initHashSet[PubSubPeer]()
+      gossipSub.topicParams[topic] = TopicParams.init()
+      gossipSub.gossipsub[topic] = initHashSet[PubSubPeer]()
+
+      for i in 0 ..< numPeers:
+        let conn = TestBufferStream.new(noop)
+        conns &= conn
+        let peerId = randomPeerId()
+        conn.peerId = peerId
+        let peer = gossipSub.getPubSubPeer(peerId)
+        peer.sendConn = conn
+        gossipSub.gossipsub[topic].incl(peer)
+
+    return (gossipSub, conns)
+
+  # Wrapper function to initialize a single topic by converting it into a seq
+  proc setupGossipSub(topic: string, numPeers: int): (TestGossipSub, seq[Connection]) =
+    setupGossipSub(@[topic], numPeers)
+
+  # Helper function to subscribe to topics
+  proc subscribeToTopics(gossipSub: TestGossipSub, topics: seq[string]) =
+    for topic in topics:
+      gossipSub.PubSub.subscribe(
+        topic,
+        proc(topic: string, data: seq[byte]): Future[void] {.async.} =
+          discard
+        ,
+      )
+
+  # Helper function to unsubscribe to topics
+  proc unsubscribeFromTopics(gossipSub: TestGossipSub, topics: seq[string]) =
+    for topic in topics:
+      gossipSub.PubSub.unsubscribeAll(topic)
+
+  # Simulate the `SUBSCRIBE` event and check proper handling in the mesh and gossipsub structures
+  asyncTest "handle SUBSCRIBE event":
```

Comment:
Not sure if using `event` is a good choice here. It has a technical meaning and I don't think that's what is happening in the spec and in the code.